File: bench/interp/bench_binarytrees.py
import sys, time
sys.setrecursionlimit(100000)
def bottom_up(d):
    if d <= 0: return (None, None)
    return (bottom_up(d-1), bottom_up(d-1))
def check(n):
    if n[0] is None: return 1
    return 1 + check(n[0]) + check(n[1])
max_depth = 16
t0 = time.perf_counter()
total = 0
d = 4
while d <= max_depth:
    iters = 1 << (max_depth - d + 4)
    for _ in range(iters): total += check(bottom_up(d))
    d += 2
print("result: %d" % total)
print("elapsed: %d ms" % int((time.perf_counter()-t0)*1000))

File: bench/interp/bench_binarytrees.js
function bottomUp(d) { if (d <= 0) return [null, null]; return [bottomUp(d-1), bottomUp(d-1)]; }
function check(n) { if (n[0] === null) return 1; return 1 + check(n[0]) + check(n[1]); }
const maxDepth = 16;
const t0 = Date.now();
let total = 0;
for (let d = 4; d <= maxDepth; d += 2) {
    const iters = 1 << (maxDepth - d + 4);
    for (let i = 0; i < iters; i++) total += check(bottomUp(d));
}
console.log("result: " + total);
console.log("elapsed: " + (Date.now() - t0) + " ms");
